### DISCUSSION TITLE
Is there a correct way to declare a class with a variable base class?

### DISCUSSION THREAD

**Discussion by btalb** [microsoft/pyright#5722] [Q&A] [answered]
I've inherited code that unfortunately does something like what's in class `C`:

```
from typing_extensions import reveal_type


class A:
    pass


class B:
    pass


if 1 + 1 == 2:
    x = A
else:
    x = B


class C(x):
    pass


y = A


class D(y):
    pass


reveal_type(x)
reveal_type(y)
```

When I run it through `pyright`, it happily takes the `D` class where the variable type isn't a union, but doesn't accept `C`:

```
$ pyright debug.py 
debug.py
  debug.py:18:9 - error: Argument to class must be a base class (reportGeneralTypeIssues)
  debug.py:29:13 - information: Type of "x" is "type[A] | type[B]"
  debug.py:30:13 - information: Type of "y" is "type[A]"
1 error, 0 warnings, 2 informations 
```

Is there a correct way to do something like this that `pyright` is happy with? Or is it just an all-round bad coding practice being correctly flagged by  `pyright`?

**Comment by erictraut** [comment, marked as answer]
> Or is it just an all-round bad coding practice being correctly flagged by pyright?

It's a very dynamic technique that can't be handled by a static type checker. The base classes of `C` obviously have a major impact on how `C` behaves (e.g. which methods it supports, etc.).

If you want your code to work well with static typing, here are a few options:

1. Define separate classes and then use a type alias to refer to the union of both.
```python
class C1(A):
    pass

class C2(B):
    pass

C = C1 | C2
```

2. If the interfaces for `A` and `B` are identical, use a cast to choose one or the other.
```python
class C(cast(type[A], x)):
    pass
```

3. Use an `Any` base class.
```python
class C(cast(Any, x)):
    pass
```

**Comment by btalb** [comment]
Thanks @erictraut . Those options are perfect.
(I'm not a fan of the original code to start with for the reasons you stated, but haven't been given a choice here)
